packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx

```diff
@@ -0,0 +1,584 @@
+    "&:hover": {
+      transform: "scale(1.2)",
+    },
+  },
+});
+
+export const MAX_THETA = 25;
+
+const isValidTarget = (e: EventTarget | null): e is HTMLDivElement =>
+  !!e && "tagName" in e && (e.tagName === "DIV" || e.tagName === "HEADER");
+
+const emojisQuery = gql`
+  query BannerEmojis {
+    BannerEmojis {
+      userId
+      displayName
+      emoji
+      link
+      description
+      x
+      y
+      theta
+    }
+  }
+`;
+
+const addEmojiMutation = gql`
+  mutation AddBannerEmoji(
+    $emoji: String!,
+    $link: String!,
+    $description: String!,
+    $x: Float!,
+    $y: Float!,
+    $theta: Float!
+  ) {
+    AddBannerEmoji(
+      emoji: $emoji,
+      link: $link,
+      description: $description,
+      x: $x,
+      y: $y,
+      theta: $theta
+    ) {
+      userId
+      displayName
+      emoji
+      x
+      y
+      theta
+    }
+  }
+`;
+
+const removeEmojiMutation = gql`
+  mutation RemoveBannerEmoji($userId: String!) {
+    RemoveBannerEmoji(userId: $userId) {
+      userId
+      displayName
+      emoji
+      link
+      description
+      x
+      y
+      theta
+    }
+  }
+`;
+
+export type BannerEmoji = {
+  userId: string,
+  displayName: string,
+  emoji: string,
+  x: number,
+  y: number,
+  theta: number,
+  link: string,
+  description: string,
+}
+
+const emojiContext = createContext<{
+  currentUser: UsersCurrent | null,
+  setEmojis: (emojis: BannerEmoji[]) => void,
+  addEmoji: (
+    emoji: string,
+    link: string,
+    description: string,
+    x: number,
+    y: number,
+    theta: number,
+  ) => Promise<{data?: {AddBannerEmoji: BannerEmoji[]}}>,
+  removeEmoji: (userId: string) => Promise<void>,
+  insertEmoji: string,
+  setInsertEmoji: (value: string) => void,
+  onCancelInsert: () => void,
+  classes: ClassesType<typeof styles>,
+} | null>(null);
+
+const useEmojiContext = () => {
+  const value = useContext(emojiContext);
+  if (!value) {
+    throw new Error("No emoji context provider");
+  }
+  return value;
+}
+
+const EmojiPicker = ({onChange}: {onChange?: (value: string) => void}) => {
+  const ref = useRef<HTMLElement | null>(null);
+  useEffect(() => {
+    const picker = ref.current;
+    if (picker) {
+      const styleId = "fm-emoji-style";
+      const styleElem = picker.shadowRoot?.getElementById(styleId);
+      if (!styleElem) {
+        const style = document.createElement("style");
+        style.id = styleId;
+        style.textContent = `input{font-family:"Inter";}`;
+        picker.shadowRoot?.appendChild(style);
+      }
+
+      const onChangeHandler = (ev: AnyBecauseTodo) => {
+        const emoji = ev?.detail?.unicode;
+        if (emoji && typeof emoji === "string") {
+          onChange?.(emoji);
+        }
+      }
+      picker.addEventListener("emoji-click", onChangeHandler);
+      return () => picker?.removeEventListener("emoji-click", onChangeHandler);
+    }
+  }, [ref, onChange]);
+  return (
+    // @ts-ignore
+    <emoji-picker ref={ref} />
```

Comment:
There's probably a simple way to augment the interface to make this typecheck, but this is urgent so I didn't bother figuring it out.